packages/telemetry/src/telemetryReporter.ts

```diff
@@ -69,7 +69,78 @@ export class TelemetryReporter extends AsyncCreatable<TelemetryOptions> {
         throw new SfdxError(messages.getMessage('unknownError'), 'unknownError', undefined, undefined, e);
       }
     } else {
-      this.logger.warn('Failed to send telemetry event because the appInsightsClient does not exist');
+      this.logger.warn('Failed to send telemetry data because the appInsightsClient does not exist');
+      throw SfdxError.create('@salesforce/telemetry', 'telemetry', 'sendFailed');
+    }
+  }
+
+  /**
+   * Publishes exception to app insights dashboard
+   * @param exception {Error} - exception you want published.
+   * @param attributes {Attributes} - map of measurements to publish alongside the exception.
+   */
+  public sendTelemetryException(exception: Error, attributes: Attributes = {}): void {
```

Comment:
There's a lot of repeated code among all these new methods. I'd love to see some sort of method that does all the disabled logic, try/catch logic, etc.. and accepts the app insights function as an argument.

For example:

```typescript
  public sendTelemetry(method: TelemetryMethod, message: string, properties?: Properties): void {
    if (!isSfdxTelemetryEnabled(this.env)) return;

    if (this.appInsightsClient) {
      this.logger.debug(`Sending telemetry: ${message}`);
      try {
        this.appInsightsClient[method]({ message, properties });
        this.appInsightsClient.flush();
      } catch (e) {
        const messages = Messages.loadMessages('@salesforce/telemetry', 'telemetry');
        throw new SfdxError(messages.getMessage('unknownError'), 'unknownError', undefined, undefined, e);
      }
    } else {
      this.logger.warn('Failed to send telemetry data because the appInsightsClient does not exist');
      throw SfdxError.create('@salesforce/telemetry', 'telemetry', 'sendFailed');
    }
  }
```

Reply:
Since there are some variations on the required object members of each type of telemetry data, we probably can't get quite that elegant in Typescript without running afoul of typing issues, but I'll give it a go.

Reply:
@mdonnalley Take a look at the changes in `fix: refactor common logic to new function`